pkg/kubelet/cm/devicemanager/topology_hints.go

```diff
@@ -179,3 +222,51 @@ func (m *ManagerImpl) getNUMANodeIds(topology *pluginapi.TopologyInfo) []int {
 	}
 	return ids
 }
+
+func (m *ManagerImpl) getPodDeviceRequest(pod *v1.Pod) map[string]int {
+	// compute the max number of resource request
+	initContainerResources := make(map[string]int)
+
+	for _, container := range pod.Spec.InitContainers {
+		for resourceObj, requestedObj := range container.Resources.Limits {
+			resource := string(resourceObj)
+			requested := int(requestedObj.Value())
+
+			if m.isDevicePluginResource(resource) {
+				if stored, ok := initContainerResources[resource]; ok {
+					if requested > stored {
+						initContainerResources[resource] = requested
+					}
+				} else {
+					initContainerResources[resource] = requested
+				}
+			}
+		}
+	}
+
+	// compute the sum of device request of user container
+	userContainerResources := make(map[string]int)
+	for _, container := range pod.Spec.Containers {
+		for resourceObj, requestedObj := range container.Resources.Limits {
+			resource := string(resourceObj)
+			requested := int(requestedObj.Value())
+
+			if m.isDevicePluginResource(resource) {
+				userContainerResources[resource] += requested
+			}
+		}
+	}
+
+	// compare them and take the bigger number
+	for userResource, userRequested := range userContainerResources {
+		if initRequested, ok := initContainerResources[userResource]; ok {
+			if userRequested > initRequested {
+				initContainerResources[userResource] = userRequested
+			}
+		} else {
+			initContainerResources[userResource] = userRequested
+		}
+	}
+
+	return initContainerResources
```

Comment:
I would reimplement this function as follows to make it more readable (note the change in name from `user` container to `app` container since this is the official name for them).
```
func (m *ManagerImpl) getPodDeviceRequest(pod *v1.Pod) map[string]int {
    podResources := sets.NewString()

    // Find the max request of a given resource across all init containers
    initContainerRequests := make(map[string]int)
    for _, container := range pod.Spec.InitContainers {
        for resourceObj, requestedObj := range container.Resources.Limits {
            resource := string(resourceObj)
            requested := int(requestedObj.Value())

            if !m.isDevicePluginResource(resource) {
                continue
            }

            podResources.Insert(resource)

            if _, exists := initContainerRequests[resource]; !exists {
                initContainerRequests[resource] = requested
                continue
            }
            if requested > initContainerRequests[resource] {
                initContainerRequests[resource] = requested

            }
        }
    }

    // Compute the sum of requests across all app containers for a given resource
    appContainerRequests := make(map[string]int)
    for _, container := range pod.Spec.Containers {
        for resourceObj, requestedObj := range container.Resources.Limits {
            resource := string(resourceObj)
            requested := int(requestedObj.Value())

            if !m.isDevicePluginResource(resource) {
                continue
            }
            podResources.Insert(resource)
            appContainerRequests[resource] += requested
        }
    }

    // Calculate podRequests as the max of init and app container requests for a given resource
    podRequests := make(map[string]int)
    for resource := range podResources {
        _, initExists := initContainerRequests[resource]
        _, appExists := appContainerRequests[resource]

        if initExists && !appExists {
            podRequests[resource] = initContainerRequests[resource]
            continue
        }

        if !initExists && appExists {
            podRequests[resource] = appContainerRequests[resource]
            continue
        }

        if initContainerRequests[resource] > appContainerRequests[resource] {
            podRequests[resource] = initContainerRequests[resource]
            continue
        }

        podRequests[resource] = appContainerRequests[resource]
    }

    return podRequests
}
```

Reply:
Thanks for providing this! Applied to the code.